File: Creational_Patterns/Factory_Method/FactoryMethod.java
// Factory Pattern
interface Shape {
    void draw();
}

class Circle implements Shape {
    @Override
    public void draw() {
        System.out.println("Shape is Circle");
    }
}

class Rectangle implements Shape {
    @Override
    public void draw() {
        System.out.println("Shape is Rectangle");
    }
}

class ShapeFactory {
    Shape getShape(String strInput) {
        if (strInput.equals("Circle")) {
            return new Circle();
        } else if (strInput.equals("Rectangle")) {
            return new Rectangle();
        } else {
            return null;
        }
    }
}

public class Main {
    public static void main(String[] args) {
        ShapeFactory shapeFactoryObj = new ShapeFactory();
        Shape rectangle = shapeFactoryObj.getShape("Circle");
        rectangle.draw();
    }
}

File: Creational_Patterns/Factory_Method/FactoryMethod.cpp
#include <bits/stdc++.h>
using namespace std;

// Factory Pattern
class Shape
{
public:
    virtual void draw() = 0;
};

class Circle : public Shape
{
public:
    void draw() override
    {
        cout << "Shape is Circle" << endl;
    }
};

class Rectangle : public Shape
{
public:
    void draw() override
    {
        cout << "Shape is Rectangle" << endl;
    }
};

class ShapeFactory
{
public:
    Shape *getShape(string strInput)
    {
        if (strInput == "Circle")
        {
            return new Circle();
        }
        else if (strInput == "Rectangle")
        {
            return new Rectangle();
        }
        else
        {
            return nullptr;
        }
    }
};

int main()
{
    ShapeFactory *shapeFactoryObj = new ShapeFactory();
    Shape *rectangle = shapeFactoryObj->getShape("Circle");
    rectangle->draw();
    return 0;
}